### PR TITLE
Simplify configs.setup

### PR DESCRIPTION
**Pull request by theHamsta**
This PR tries to simplify `configs.setup` to avoid a few problems. The main problem: attach/detach should be strictly inverse to each other and if they are not it shouldn't cause a problem for disabled modules.

- try to avoid enabling modules just to disable them afterwards for languages in `disable`. This could avoid problems when attach/detach of certain modules are not inverse to each other
- make attach/detach of `highlight_current_scope` inverse to each other
- avoid calling `enable_all` with disable/enable data that is actually for a different module

This solves the bug that executing this lua file (once all modules are loaded) automatically activates `highlight_current_scope` though it's deactivated. Happens with every top-level module set to true.

```lua
    require "nvim-treesitter.configs".setup(
        {
            highlight = {
                enable = true, -- false will disable the whole extension
                disable = {"html", "lua"} -- list of language that will be disabled
            },
            refactor = {
                highlight_current_scope = {
                    enable = false,
                    inverse_highlighting = true,
                    disable = {"python", "markdown"}
                },
                highlight_definitions = {
                    enable = true,
                    disable = {"markdown"}
                },
            },
            ensure_installed = "all",
        }
    )
```

This is still a draft. There were a few bugs that I discovered on the way. I found the function how it was before very confusing and I could not understand it fully. So please handle those changes with care.

### CODE REVIEW

**Comment by vigoux** [comment]
Is that really a draft ?

File: lua/nvim-treesitter/configs.lua [hunk 1]
```diff
@@ -295,24 +299,17 @@ end
...
-      recurse_modules(function(_, _, new_path)
-        if data.enable then
-          enable_all(new_path)
-        end
-
-        for _, lang in ipairs(data.disable or {}) do
-          disable_mod_conf_autocmd(new_path, lang)
-        end
-      end, config.modules)
+  config.modules = vim.tbl_deep_extend('force', config.modules, user_data)
```
Comment (theHamsta) [inline lua/nvim-treesitter/configs.lua:302]: Why do we have modules here? This causes our table structure to be slightly different from the one the user provides us and it's only for one boolean value.

Comment (vigoux) [reply lua/nvim-treesitter/configs.lua:302]: We need that because of the use of the recurse module below

File: lua/nvim-treesitter/configs.lua [hunk 2]
```diff
@@ -295,24 +299,17 @@ end
...
+
+  config.ensure_installed = user_data.ensure_installed
```
Comment (vigoux) [inline lua/nvim-treesitter/configs.lua:304]: I don't think we need the line anymore

File: lua/nvim-treesitter/configs.lua [hunk 3]
```diff
@@ -295,24 +299,17 @@ end
...
+  require'nvim-treesitter.install'.ensure_installed(config.ensure_installed)
```
Comment (theHamsta) [inline lua/nvim-treesitter/configs.lua:305]: We could also just not store `ensure_installed`. We don't need it any more. Actually, I should also delete `enure_installed` from `config.modules`.

File: lua/nvim-treesitter/refactor/highlight_current_scope.lua [hunk 4]
```diff
@@ -37,8 +37,8 @@ function M.attach(bufnr)
 end
 
 function M.detach(bufnr)
-  M.clear_usage_highlights(bufnr)
```
Comment (theHamsta) [inline lua/nvim-treesitter/refactor/highlight_current_scope.lua:40]: This function does not exist anymore. Lol.

File: lua/nvim-treesitter/configs.lua (hunk 1 above)
Comment (theHamsta) [reply lua/nvim-treesitter/configs.lua:302]: Recurse modules does not care about a boolean because it will do a `is_module` check.

Comment (vigoux) [reply lua/nvim-treesitter/configs.lua:302]: Then you can even remove the boolean afterwards to save us some time

**Comment by theHamsta** [comment]
Would be good to have @steelsojka a look on this. I only know slightly how the module system works and I really don't want to break something.

**Comment by theHamsta** [comment]
The blocklist does not work yet... :-(

**Review by vigoux — COMMENTED** [review]
Please just remove the `ensure_installed` key from `config.modules` and we will be good to go.

**Review by vigoux — APPROVED** [review]

**Comment by theHamsta** [comment]
Add this?
```lua
if user_data.ensure_installed then
  require'nvim-treesitter.install'.ensure_installed(user_data.ensure_installed)
  config.modules.ensure_installed = nil
end
```
`ensure_installed` will also work with nil but the require would not be executed.

**Review by vigoux — APPROVED** [review]

**Pull request merged**